lms/config/resources.py

```diff
@@ -35,11 +34,17 @@ class LTILaunch:
     def __init__(self, request):
         """Return the context resource for an LTI launch request."""
         self._request = request
+        self._authority = self._request.registry.settings["h_authority"]
+        self._h_pubid = None
+
         # This will raise HTTPBadRequest if the request looks like an OAuth
         # redirect request but no DB-stashed LTI params can be found for the
         # request.
         self._lti_params = lti_params_for(request)
-        self._authority = self._request.registry.settings["h_authority"]
+
+        if "pubid" in self._lti_params:
+            pubid = self._lti_params.pop("pubid")
+            self._h_pubid = pubid
```

Comment:
This is absolutely horrible but I think it's all we can do for now.

In the case of a Canvas file assignment launch (and only Canvas files assignments, this doesn't apply to any other kind of assignment) the app does this:

1. Receives an LTI launch request from Canvas.
2. The various h API requests to create the user, create or update the group, add the user to the group as a member, and set `context.pubid` to the `id` received from h, all happen as normal.
3. Saves the LTI launch params in the database.
4. Redirects the iframe to Canvas's OAuth authorization page. So `context.pubid` was never used, and is now lost.
5. User clicks <kbd>Authorize</kbd>.
6. Canvas redirects the iframe back to our OAuth redirect URL.
7. The app now actually launches the assignment in response to this OAuth redirect request from Canvas, _not_ in response to the original LTI launch request (step 1) itself. The app retrieves the launch params that it stashed in the DB in step 3 in order to do the launch. But at this point in time `context.pubid` is `None`. It was set during request 1 (the LTI launch request) but has not yet been set during request 6 (the OAuth redirect request).

I've hacked the code to also stash the group's `pubid` along with the launch params (see below). These three lines here restore that stashed `pubid`.

Reply:
So the crux of the issue is that in the Canvas-file launch, the launch is handled over multiple requests instead of one request, so context is lost? Correct?

Reply:
It's so icky that I wonder if you might add a TODO/FIXME comment here?

Reply:
> So the crux of the issue is that in the Canvas-file launch, the launch is handled over multiple requests instead of one request, so context is lost? Correct?

Yep. It does this for all Canvas file assignment launches and all content item selection launches within Canvas.

> It's so icky that I wonder if you might add a TODO/FIXME comment here?

I didn't do this because if I tried to comment every icky thing about the LMS app, it'd be all comments. This particular context stash and unstash hack, for example, is already used all over the place (and has had to be used by the provisioning code a few times already). I think we should just move as fast as possible towards fixing the code design including completely replacing this write-off of an OAuth implementation.

Reply:
I think Rob is gonna investigate but if the client's `groups` setting could accept `groupid`s then we could remove this hack https://hypothes-is.slack.com/archives/C4K6M7P5E/p1543925103033300